gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -3,159 +3,164 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+from itertools import product
+import warnings
 import torch
 from .lazy_tensor import LazyTensor
 
 
 class DiagLazyTensor(LazyTensor):
-    def __init__(self, diag, exact_root_decomposition=False):
+    def __init__(self, diag):
         """
-        Diagonal lazy tensor
+        Diagonal lazy tensor. Supports arbitrary batch sizes.
 
         Args:
-            - diag (tensor) (batch) diagonal of matrix
-            - exact_root_decomposition (bool, default False) return an exact root decomposition.
-                Set to True only if the represented diagonal is relatively small.
+            :attr:`diag` (Tensor):
+                A `b1 x ... x bk x n` Tensor, representing a `b1 x ... x bk`-sized batch
+                of `n x n` diagonal matrices
         """
         super(DiagLazyTensor, self).__init__(diag)
         self._diag = diag
-        self.exact_root_decomposition = exact_root_decomposition
+        self._batch_shape = diag.shape[:-1]
+        self._batch_dim = len(self._batch_shape)
 
-    def _matmul(self, rhs):
-        if rhs.ndimension() == 1 and self.ndimension() == 2:
-            return self._diag * rhs
+    def __add__(self, other):
+        if isinstance(other, DiagLazyTensor):
+            return self.add_diag(other._diag)
+        from .added_diag_lazy_tensor import AddedDiagLazyTensor
+
+        return AddedDiagLazyTensor(other, self)
+
+    def _batch_get_indices(self, batch_indices, left_indices, right_indices):
+        warnings.warn("_batch_get_indices is deprecated - use _get_indices instead", DeprecationWarning)
+        return self._get_indices(left_indices, right_indices, batch_indices)
+
+    def _get_indices(self, left_indices, right_indices, *batch_indices):
+        """Extract elements from the LazyTensor. Supports arbitrary batch sizes.
+
+        Args:
+            :attr:`left_indices` (LongTensor): An `i`-dim tensor of row indices (these are the
+                same across all batches).
+            :attr:`left_indices` (LongTensor): An `i`-dim tensor of column indices (these are the
+                same across all batches).
+            :attr:`batch_indices` (LongTensor): A variable number of `i`-dim tensors of batch indices,
+                one for each batch dimension. If smaller than the number of batch dimensions of the
+                DiagLazyTensor, select all elements of the omitted batches.
+
+        Returns:
+            An `i`-dim tensor containing the requested elements if all batch index tensors were
+            passed in, otherwise a `bl x ... bk x i` batch of such tensors, where `l = len(batch_indices) + 1`.
+
+        """
+        batch_dims = len(batch_indices)
+        if batch_dims > self._batch_dim:
```

Comment:
Since `_get_indices` is an internally used method, I don't know if this check is necessary.  Additionally, I think it should always be the case `len(batch_indices) == self._batch_dim`, unless this changed somehow with broadcasting.

Reply:
yeah this method implements "broadcasting" in the sense described in the docstring: 

> `:attr:batch_indices (LongTensor)`: A variable number of `i`-dim tensors of batch indices, one for each batch dimension. If smaller than the number of batch dimensions of the DiagLazyTensor, select all elements of the omitted batches.

Reply:
Also, even though it's internal, having good error messages makes things much easier to debug...